### ISSUE TITLE
Non-stick-aligned sequence length mis-lowers scaled_dot_product_attention score matmul

### ISSUE THREAD

**Issue by arielge** [torch-spyre/torch-spyre#3113]
## What happened?

`torch.nn.functional.scaled_dot_product_attention` produces wrong numeric results on Spyre when the sequence length is not stick-aligned. A standalone repro using only `torch` on tensors with shape `q,k,v = [1, 16, 3500, 64]` in `torch.float16` shows that the unpadded sequence length `3500` (not divisible by 64) diverges significantly from a CPU reference, while an algebraically-equivalent padded variant `3520` (right-pad Q/K/V rows, mask pad keys with `-inf`, then compare only the real rows) matches the CPU reference closely. This was localized from the Pixtral vision tower used by Mistral3 Vision. On CPU, the unpadded and padded-real outputs match to `max|base - padded_real| = 0.000061`, so the pad+mask transform is effectively a no-op for the real rows. On Spyre, the same two variants differ by `max|base - padded_real| = 0.345459`, which indicates a device-side lowering/numeric defect rather than a reference or math error.

Related to #3098 in that both are attention wrong-results bugs, but this issue is about plain SDPA on a non-stick-aligned sequence length rather than coarse `Lq` tiling in folded GQA flash attention.

## What did you expect to happen?

Spyre SDPA should match the CPU reference within a small absolute tolerance regardless of whether the mathematically-equivalent computation is run with the original sequence length `3500` or with a right-padded sequence length `3520` and pad keys masked out. In the attached repro, both Spyre variants should agree with the CPU result on the real rows within roughly `5e-2`, and the two Spyre variants should agree with each other as well.

## How can we reproduce it?

Run the following standalone script on a Spyre pod. The unpadded (`3500`) and padded (`3520`) variants have different sequence lengths, so they compile to distinct binaries — no same-shape compile-cache reuse is possible between them, and the Spyre-vs-Spyre comparison is valid within a single process.

```python
import torch
import torch.nn.functional as F

DTYPE = torch.float16
DEVICE = torch.device("spyre")
BLOCK_SIZE = 64
BATCH = 1
HEADS = 16
SEQ = 3500
PADDED_SEQ = 3520
HEAD_DIM = 64
SCALE = HEAD_DIM**-0.5
CPU_ATOL = 5e-3
SPYRE_ATOL = 5e-2
SEED = 1234


def make_inputs():
    torch.manual_seed(SEED)
    q = torch.randn(BATCH, HEADS, SEQ, HEAD_DIM, dtype=DTYPE)
    k = torch.randn(BATCH, HEADS, SEQ, HEAD_DIM, dtype=DTYPE)
    v = torch.randn(BATCH, HEADS, SEQ, HEAD_DIM, dtype=DTYPE)
    return q, k, v


def right_pad_qkv(q, k, v, padded_seq):
    pad = padded_seq - q.shape[2]
    zeros = lambda t: torch.zeros(*t.shape[:2], pad, t.shape[3], dtype=t.dtype)
    return (
        torch.cat([q, zeros(q)], dim=2),
        torch.cat([k, zeros(k)], dim=2),
        torch.cat([v, zeros(v)], dim=2),
    )


def build_pad_key_mask(seq, padded_seq, dtype):
    mask = torch.zeros(1, 1, padded_seq, padded_seq, dtype=dtype)
    if padded_seq > seq:
        mask[:, :, :, seq:] = float("-inf")
    return mask


def sdpa(q, k, v, attn_mask):
    return F.scaled_dot_product_attention(
        q, k, v, attn_mask=attn_mask, dropout_p=0.0, is_causal=False, scale=SCALE
    )

q, k, v = make_inputs()
q_pad, k_pad, v_pad = right_pad_qkv(q, k, v, PADDED_SEQ)
pad_mask = build_pad_key_mask(SEQ, PADDED_SEQ, DTYPE)

with torch.no_grad():
    cpu_base = sdpa(q.float(), k.float(), v.float(), None).to(DTYPE)
    cpu_pad = sdpa(q_pad.float(), k_pad.float(), v_pad.float(), pad_mask.float()).to(DTYPE)
    cpu_pad_real = cpu_pad[:, :, :SEQ, :]
    print("CPU control max|base - padded_real| =", (cpu_base - cpu_pad_real).abs().max().item())

with torch.no_grad():
    spyre_base = torch.compile(sdpa, dynamic=False)(q.to(DEVICE), k.to(DEVICE), v.to(DEVICE), None).to("cpu")
    spyre_pad = torch.compile(sdpa, dynamic=False)(
        q_pad.to(DEVICE), k_pad.to(DEVICE), v_pad.to(DEVICE), pad_mask.to(DEVICE)
    ).to("cpu")
    spyre_pad_real = spyre_pad[:, :, :SEQ, :]

print("Spyre max|base - cpu| =", (spyre_base - cpu_base).abs().max().item())
print("Spyre max|padded_real - cpu| =", (spyre_pad_real - cpu_base).abs().max().item())
print("Spyre max|base - padded_real| =", (spyre_base - spyre_pad_real).abs().max().item())
```

Observed output from an actual Spyre run:

```text
CPU control max|base - padded_real| = 0.000061
Spyre max|base - cpu|       = 0.344727
Spyre max|padded_real - cpu|= 0.002563
Spyre max|base - padded_real| = 0.345459
VERDICT: FAIL reproduced
```

## Any environmental details we need to know?

- torch `2.11.0+cu130`
- torch-spyre `0.0.1`
- torch-spyre commit `7e45168f1d56ca1cec4889a3e19b14719dcdd23f`
- dtype: `torch.float16`
- device: Spyre pod (`torch.device("spyre")`)

## Anything else we need to know?

This was first observed while bringing up the Pixtral vision tower in [`hf_adapters/hf_pixtral_vision.py`](hf_adapters/hf_pixtral_vision.py) and the full Mistral3 multimodal adapter in [`hf_adapters/hf_mistral3_vision_mm.py`](hf_adapters/hf_mistral3_vision_mm.py). In that model path, the real patch sequence length is `P=3500`, and right-padding the patch sequence to a `BLOCK_SIZE` multiple before SDPA is the current workaround.

## Relevant log output

```Shell
CPU control max|base - padded_real| = 0.000061
Spyre max|base - cpu|       = 0.344727
Spyre max|padded_real - cpu|= 0.002563
Spyre max|base - padded_real| = 0.345459
VERDICT: FAIL reproduced
```

**Comment by arielge** [comment]
## Additional data point: the same non-stick-aligned `P` also makes a **RoPE→SDPA fusion overflow fp16 → `Inf`** (not just wrong-but-finite)

While bringing up the Pixtral vision tower (Mistral-Small-3.1-24B) where this issue's ragged-`P` defect was found, we hit a **second, distinct** symptom at the *same* non-stick-aligned sequence length `P = 3500` (`3500 % 64 == 44`). Filing it here since it is triggered by the identical geometry and is masked by the same stick-pad workaround.

**Symptom:** when a matmul-form RoPE is fused into the *same compiled graph* as `scaled_dot_product_attention`, the fused SDPA output **overflows fp16 → `Inf`** — even though the *standalone* SDPA on the identical inputs is finite, and the CPU-fp32 reference is finite and modest-magnitude. Inserting a `.contiguous()` between the RoPE and the SDPA — a provable no-op in exact arithmetic — breaks the fusion and restores finiteness. So this is a lowering-*selection* defect, not real fp16 range exhaustion.

Relationship to this issue: the plain score matmul at a ragged `P` is *wrong-but-finite* (the top post), and *additionally* the fused RoPE→SDPA graph goes all the way to `Inf`. Both disappear once `P` is padded to a `BLOCK_SIZE` multiple — so the same tiling fix very likely resolves both. Flagging the fused→`Inf` path so it's covered by the fix and its regression test.

### Standalone repro

Seeded random q/k/v with the Pixtral geometry (`[1, 16, 3500, 128]`, `scale = 1/√64`). The fused-only `Inf` is magnitude-sensitive (it is a lowering choice); `--qk-scale 3.5` (the default) is the sweet spot where **standalone stays finite but the fused graph overflows** — isolating the fusion, not just pushing everything over. Run one variant per process (airtight against compile-cache reuse):

```bash
for V in standalone fused contiguous; do
  python3 repro_rope_sdpa_fusion_inf_spyre.py --variant $V
done
```

```python
import argparse

import torch
import torch.nn.functional as F

DTYPE = torch.float16
DEVICE = torch.device("spyre")

# Pixtral vision tower geometry (Mistral-Small-3.1-24B, last block).
BATCH, HEADS, SEQ, HEAD_DIM = 1, 16, 3500, 128  # SEQ % 64 == 44 -> NOT stick-aligned
ROPE_HALF, ORIG_HALF = 64, 32
SCALE = 64 ** -0.5
SEED = 0
QK_SCALE = 3.5  # standalone finite, fused Inf; <=3.0 all finite, >=4.0 standalone also Inf


def apply_rope_matmul(x, rope):  # from hf_adapters/hf_common.py: RoPE as matmul, no slicing
    B, H, L, D = x.shape
    x_ = x.transpose(1, 2).reshape(B, L, H, 2, D // 2)
    return rope[:, :, None].mul(x_.unsqueeze(-3)).sum(4, keepdim=True).flatten(3).transpose(1, 2)


def build_rope():
    pos = torch.arange(SEQ, dtype=torch.float32)[:, None]
    inv_freq = 1.0 / (10000.0 ** (torch.arange(ORIG_HALF).float() / ORIG_HALF))
    ang = pos * inv_freq[None, :]
    rot = torch.stack([ang.cos(), -ang.sin(), ang.sin(), ang.cos()], 1).view(SEQ, 2, 2, ORIG_HALF)
    ident = torch.zeros(SEQ, 2, 2, ROPE_HALF - ORIG_HALF)
    ident[:, 0, 0, :] = ident[:, 1, 1, :] = 1.0  # identity in the padded lanes
    return torch.cat([rot, ident], -1).contiguous().to(DTYPE)[None]


def make_inputs():
    torch.manual_seed(SEED)

    def one(scale):
        t = torch.zeros(BATCH, HEADS, SEQ, HEAD_DIM, dtype=DTYPE)
        t[..., 0:ORIG_HALF] = (torch.randn(BATCH, HEADS, SEQ, ORIG_HALF) * scale).to(DTYPE)
        t[..., ROPE_HALF:ROPE_HALF + ORIG_HALF] = (torch.randn(BATCH, HEADS, SEQ, ORIG_HALF) * scale).to(DTYPE)
        return t

    return one(QK_SCALE), one(QK_SCALE), one(1.0)


def sdpa(q, k, v):
    return F.scaled_dot_product_attention(q, k, v, attn_mask=None, dropout_p=0.0, is_causal=False, scale=SCALE)


def graph(variant):
    if variant == "standalone":  # RoPE applied outside the compiled graph
        return lambda q, k, v, r: sdpa(q, k, v)
    if variant == "fused":  # RoPE matmul fused into the SDPA graph
        return lambda q, k, v, r: sdpa(apply_rope_matmul(q, r), apply_rope_matmul(k, r), v)
    if variant == "contiguous":  # .contiguous() buffer boundary between RoPE and SDPA
        return lambda q, k, v, r: sdpa(apply_rope_matmul(q, r).contiguous(), apply_rope_matmul(k, r).contiguous(), v)
    raise ValueError(variant)


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--variant", required=True, choices=["standalone", "fused", "contiguous"])
    args = ap.parse_args()

    rope = build_rope()
    q, k, v = make_inputs()
    with torch.no_grad():
        cpu_ref = sdpa(apply_rope_matmul(q.float(), rope.float()), apply_rope_matmul(k.float(), rope.float()), v.float())
    print(f"CPU ref: finite={torch.isfinite(cpu_ref).all().item()} max_abs={cpu_ref.abs().max().item():.4f}")

    compiled = torch.compile(graph(args.variant), dynamic=False)
    # standalone gets q/k already rotated so its inputs match the fused variants' post-RoPE q/k
    if args.variant == "standalone":
        q, k = apply_rope_matmul(q, rope), apply_rope_matmul(k, rope)
    with torch.no_grad():
        out = compiled(q.to(DEVICE), k.to(DEVICE), v.to(DEVICE), rope.to(DEVICE)).to("cpu").float()
    print(f"[spyre:{args.variant}] has_inf={torch.isinf(out).any().item()} max_abs={out.abs().max().item():.4f}")


if __name__ == "__main__":
    main()
```

### Observed on Spyre (verbatim, one process per variant)

```text
############## standalone ##############
CPU ref: finite=True max_abs=5.1746
[spyre:standalone] has_inf=False max_abs=5.1797

############## fused ##############
CPU ref: finite=True max_abs=5.1746
[spyre:fused] has_inf=True max_abs=inf

############## contiguous ##############
CPU ref: finite=True max_abs=5.1746
[spyre:contiguous] has_inf=False max_abs=5.1797
```

`standalone` (RoPE applied outside the compiled graph) is finite. `fused` (RoPE matmul inside the SDPA graph) overflows to `Inf`. `contiguous` (a `.contiguous()` buffer boundary between RoPE and SDPA) is finite again. Same inputs, same shapes → the `Inf` is purely a property of the fused lowering at the ragged `P`.

### Corroboration on the real model activations

The synthetic scale-3.5 case mirrors the real tower. Capturing the real block-23 post-RoPE `q/k/v` at the op boundary and running the same A/B/C/D variants (one process each):

| variant | graph | **padded `P = 3520`** | **unpadded `P = 3500`** |
|---|---|---|---|
| A | `sdpa(q_rope, k_rope, v)` standalone | finite, `max\|spy-cpu\| = 0.12` | finite but **wrong**, `max\|spy-cpu\| = 10.7` (this issue) |
| B | `sdpa(rope(q), rope(k), v)` **fused** | finite (`0.17`) | **`Inf`** |
| C | `sdpa(rope(q).contiguous(), rope(k).contiguous(), v)` | finite (`0.17`) | **finite** (`= A`) |
| D | `sdpa(q_rope.contiguous(), k_rope.contiguous(), v)` | finite (`0.12`) | finite (`= A`) |

The real activations overflow at their native magnitude (`q max_abs ≈ 12`) because their attention is peaked on structured directions; the synthetic repro reaches the same regime with `--qk-scale 3.5`. Crucially, at **padded `P = 3520` the fused variant does NOT overflow** — so this `Inf`, like the wrong-numbers score matmul, is gated entirely by the non-stick-aligned `P`.

### Environment

- torch `2.11.0+cu130`
- torch-spyre `0.0.1`, commit `7e45168f1d56ca1cec4889a3e19b14719dcdd23f`
- device: Spyre pod (`torch.device("spyre")`), AIU pf, dtype `torch.float16`
